external/qjs_query_string_module.c

```diff
@@ -0,0 +1,1010 @@
+static JSValue qjs_query_string_stringify(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static JSValue qjs_query_string_escape(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static JSValue qjs_query_string_unescape(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static JSValue qjs_query_string_decode(JSContext *cx, const u_char *start,
+    size_t size);
+static int qjs_query_string_encode(njs_chb_t *chain, njs_str_t *str);
+static JSValue qjs_query_string_parser(JSContext *cx, u_char *query,
+    u_char *end, njs_str_t *sep, njs_str_t *eq, JSValue decode,
+    unsigned max_keys);
+static JSModuleDef *qjs_querystring_init(JSContext *ctx, const char *name);
+
+
+static const JSCFunctionListEntry  qjs_querystring_export[] = {
+    JS_CFUNC_DEF("decode", 4, qjs_query_string_parse),
+    JS_CFUNC_DEF("encode", 4, qjs_query_string_stringify),
+    JS_CFUNC_DEF("escape", 1, qjs_query_string_escape),
+    JS_CFUNC_DEF("parse", 4, qjs_query_string_parse),
+    JS_CFUNC_DEF("stringify", 4, qjs_query_string_stringify),
+    JS_CFUNC_DEF("unescape", 1, qjs_query_string_unescape),
+};
+
+
+qjs_module_t  qjs_query_string_module = {
+    .name = "querystring",
+    .init = qjs_querystring_init,
+};
+
+
+static JSValue
+qjs_query_string_parse(JSContext *cx, JSValueConst this_val, int argc,
+    JSValueConst *argv)
+{
+    int64_t    max_keys;
+    JSValue    options, ret, decode, native;
+    njs_str_t  str, sep, eq;
+
+    sep.start = NULL;
+    eq.start = NULL;
+    str.start = NULL;
+
+    max_keys = 1000;
+    decode = JS_UNDEFINED;
+
+    if (!JS_IsNullOrUndefined(argv[1])) {
+        sep.start = (u_char *) JS_ToCStringLen(cx, &sep.length, argv[1]);
+        if (sep.start == NULL) {
+            return JS_EXCEPTION;
+        }
+    }
+
+    if (!JS_IsNullOrUndefined(argv[2])) {
+        eq.start = (u_char *) JS_ToCStringLen(cx, &eq.length, argv[2]);
+        if (eq.start == NULL) {
+            JS_FreeCString(cx, (char *) sep.start);
+            return JS_EXCEPTION;
+        }
+    }
+
+    options = argv[3];
+    if (JS_IsObject(options)) {
+        ret = JS_GetPropertyStr(cx, options, "maxKeys");
+        if (JS_IsException(ret)) {
+            goto fail;
+        }
+
+        if (!JS_IsUndefined(ret)) {
+            if (JS_ToInt64(cx, &max_keys, ret) < 0) {
+                JS_FreeValue(cx, ret);
+                goto fail;
+            }
+
+            JS_FreeValue(cx, ret);
+
+            if (max_keys < 0) {
+                max_keys = INT64_MAX;
+            }
+        }
+
+        decode = JS_GetPropertyStr(cx, options, "decodeURIComponent");
+        if (JS_IsException(decode)) {
+            goto fail;
+        }
+
+        if (!JS_IsUndefined(decode) && !JS_IsFunction(cx, decode)) {
+            JS_ThrowTypeError(cx, "option decodeURIComponent is not "
+                              "a function");
+            JS_FreeValue(cx, decode);
```

Comment:
Forgot to remove this `JS_FreeValue()`, but there is no issue in the latter function `qjs_query_string_stringify()`.